### ISSUE TITLE
Client-generation: Allow data validation on the client side (using minLength/maxLength)

### ISSUE THREAD

**Issue by jfiala** [swagger-api/swagger-codegen#2549]
Currently the generated Java-clients perform required-validation automatically. 
Do you think this makes sense for minLength/maxLength as well, to reject the request on the client side before sending to the server?

For Java the JSR-303 Bean Validation API could be used for this automatically.
For other languages maybe a JSON-schema validation could be used.

**Comment by wing328** [comment]
@jfiala yes, it's definitely something we want to leverage and I'm sure the community would appreciate your contribution on that.

Let me know if you need help.

Ref: https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md

**Comment by wing328** [comment]
@jfiala may I know if you need help adding the JSR-303 Bean validation to the Java API client?

**Comment by jfiala** [comment]
@wing328 thank you, maybe we can sketch some of the main features:
* I think it should be configurable so users can turn it on/off if they don't want bean validation annotations in the client code
* Bean Validation would make sense for every Java generator, but with the current layout (everything in mustache templates) it means I'd have to add it to all templates, do you have any better ideas?
* More general: Currently there are generators for server/client code kept separately, but shouldn't they be provided together (as for CXF for example where one can switch to generate impl/interface) as the model is always shared between server/client and should be the same to the bigger extent?

**Comment by jfiala** [comment]
Currently we could add support for the following annotations:
http://docs.oracle.com/javaee/6/tutorial/doc/gircz.html

* @NotNull (required=true)
* @Null (required=false)
* @Size (minLength/maxLength)
* @Min (min)
* @Max (min)
* @Pattern (pattern)
* @DecimalMin (min for BigDecimal)
* @DecimalMax (max for BigDecimal)

Pls let me know if other annotations make sense in the scope of Swagger?

**Comment by wing328** [comment]
@jfiala I think your list covers what we've already covered for other generators (e.g. PHP, Ruby).

> Bean Validation would make sense for every Java generator, but with the current layout (everything in mustache templates) it means I'd have to add it to all templates, do you have any better ideas?

Please add it to Java okhttp-gson API client to begin with and we'll do the same for other Java client/server generator one by one.

**Comment by cbornet** [comment]
Duplicate #2412 ?
See also #2663.

Also Java client pojo.mustache only differ by a few annotations between gson and jackson (@Serialized vs @JsonProperty) so it should be very easy to mutualize them under a parent mustache with a "gson" tag.

**Comment by jfiala** [comment]
@wing328: I added the first annotations to the Java language, see https://github.com/jfiala/swagger-codegen/commit/08b3b53f0ccaddaa177c242ba7fde0e0a64f3f36

Supported beanvalidation annotations (http://docs.oracle.com/javaee/6/tutorial/doc/gircz.html): 
* Null, NotNull
* Pattern, Size for string
* Min/Max for integer

Questions:
* Should we add a flag to the CodegenParameter to switch Bean validation API support on/off?
* Min/Max need an integer value, but {{minimum}}/{{maximum}} inserts a double. Should we add minimumLong/maximumLong to CodegenParameter to insert a long value?
* I added Bean validation API 1.1.0.Final to the pom.mustache, or should I use ${validation-api-version}?
* to also support BigDecimal DecimalMin/DecimalMax I need to check if the type is "number", should we add CodegenParameter.isNumber?
* As Pattern/Size only applies to strings, I'd need CodegenParameter.isString to check.
* As Min/Max only applies to integers, I'd need CodegenParameter.isInteger to check.

**Comment by cbornet** [comment]
Note : "required=false" is not equivalent to @Null

**Comment by jfiala** [comment]
The following Bean-Validation-Annotations have not been implemented because there seems no Swagger-spec equivalent (pls let me know if there is...):
* @AssertTrue/@AssertFalse = he value of the field or property must be true/false.
* @Digits = The value of the field or property must be a number within a specified range. The integer element specifies the maximum integral digits for the number, and the fraction element specifies the maximum fractional digits for the number.
* @Null = The value of the field or property must be null.
* @Past/@Future = The value of the field or property must be a date in the past/future.

**Comment by wing328** [comment]
For Min/Max of Integer/Long/Decimal, I'm working on a PR so that we only need 2 fields to present any min/max value. I'll share more update shortly.

**Comment by jfiala** [comment]
@wing328: I put the annotations on the getter for now - just wondering if it is better for readability to put it at the field.
When setting up a model I'm usually putting the annotations on the fields and I'd like to re-generate my model from the swagger-spec... However, regarding this http://stackoverflow.com/questions/6283726/jsf-jsr-303-bean-validation-why-on-getter-and-not-setter it might be better to leave it at the getter. 
Any thoughts?

**Comment by jfiala** [comment]
@wing328 I'd like to add some docs about the beanvalidation (supported annotations etc.), where shall I put this?
Additionally, I think it would be nice to have some docs for each language explaining the features supported etc.?

**Comment by cbornet** [comment]
You can use the README for that.

**Comment by jfiala** [comment]
updated the readme: https://github.com/jfiala/swagger-codegen/tree/beanvalidation_2549

**Comment by jfiala** [comment]
Open issues:
* add a switch for bean-validation
* maybe externalize the bean-validation snippet as a separate mustache file for all java clients
* add test cases for bean-validation

**Comment by brentryan** [comment]
What's the ETA on this feature?  Where can we help so we can complete this and get JSR-303 annotations in?

**Comment by jfiala** [comment]
I added bean-validation to Java and JAXRS CXF client generation. I'd like to know if it is preferred to pull bean-validation out into a separate module and reuse this across all Java clients  (and add test cases only at one place) or if it is preferred to simply copy & paste this to all Java clients (and have redundant test-cases or decide to only create test cases for e.g. the Java client).

IMHO if there is no immediate decision on this, I can fine-tune the Java/JAXRS CXF client (add switch + test cases) and then pull at least these two into the next release? Then modularization for all Java clients could be a topic for the next release...

**Comment by fehguy** [comment]
@jfiala seems like best would be if this was a library option like JAXRS (between 1.x and 2.x) , so you can enable as desired.  Does that make sense?

**Comment by jfiala** [comment]
@fehguy Yes definitely, I'll implement the switch for JAXRS CXF and let you know for review, then you can decide how to apply this to all other languages...

**Comment by jfiala** [comment]
@fehguy I implemented the switch useBeanValidation as a CLI switch / additionalProperty). Added it to the Java and CXF client (see this PR: https://github.com/swagger-api/swagger-codegen/pull/3281).

The beanvalidation switch is now implemented in JavaClientCodegen.java and JavaCXFServerCodegen.java, I'd suggest moving it upward to AbstractJavaCodegen (where it actually best fits) - after doing so, each language can add the switch individually.

What do you think? Pls let me know if you need anything else to get it in...
Pls advise where you would like to have tests added (for each language?)?

**Comment by jfiala** [comment]
@fehguy regarding the copy & paste work of the BeanValidation-annotations: What do you think about adding a preprocessing-directive which allows to reuse a mustache template (or does this already exist in the project?)?

Like this:
1. Define a Mustache file property_beanValidation.mustache
2. Define  a list of preloaded Mustache-files in the language Java
"property_beanValidation" ==> "property_beanValidation.mustache"
3. Load the mustache file inside the pojo.mustache:
...
{{#useBeanValidation}}
[[property_beanValidation]]
{{/useBeanValidation}}
....
This would allow us to keep the BeanValidation definition at a central place and reuse it for all sorts of Java-clients...
What do you think?

**Comment by jfiala** [comment]
@fehguy: added a reusable mustache subtemplate to the lanugage java.
This can be easily reused in other languages using "{{#useBeanValidation}}{{>..\..\java\beanValidation}}{{/useBeanValidation}}".
Pls let me know if you prefer a different place for this new reusable template.

Next step is to move useBeanValidation to the AbstractJavaCodegen and then the other languages only need to add the subtemplate to their pojo.mustache.
Pls give me OK to move useBeanValidation to AbstractJavaCodegen ?

**Comment by wing328** [comment]
> {{>..\..\java\beanValidation}}

Would that work in non-windows environment (mac, linux, etc) as well?

**Comment by jfiala** [comment]
Probably not, corrected...

**Comment by jfiala** [comment]
Wouldn't a centralized file for the Java language also make sense for the comments above the getters as well? Using the subtemplating we could extract all these into the Java language?

**Comment by abhijith-prabhakar** [comment]
Is this accounting for minItems and maxItems on an array property of a schema?  As far as I can see Codegen is not reading these values(CodegenProperty.java).  Its read only in CodegenParameter.java

**Comment by jfiala** [comment]
@abhijith-prabhakar The Beanvalidation mustache file is using the minLength/maxLength of to assemble @Size, minItems/maxItems is currently not used. After minItems/maxItems is available, I can add it to the mustache templates.

**Comment by jfiala** [comment]
@wing328 Currently there are language tests verifying if the CodegenModel is assembled correctly. However, I found no tests verifying the actual generator results from the mustache files (e.g. generated Java sources). Do you think it makes sense adding such tests?

**Comment by jfiala** [comment]
@wing328 Can you pull in the current implementation of the beanvalidation or are you missing something?

**Comment by wing328** [comment]
@jfiala I'll take another look later today.

**Comment by jfiala** [comment]
@abhijith-prabhakar added support for minItems/maxItems (maps to @Size).

**Comment by abhijith-prabhakar** [comment]
Thank you.  Will try it out once accepted

**Comment by wing328** [comment]
https://github.com/swagger-api/swagger-codegen/pull/4068 merged into master. Please pull the latest master to give it a try.

**Comment by jfiala** [comment]
@wing328 We now have the necessary BeanValidation-annotations at the client code. 

To perform validation automatically, we need to invoke the BeanValidation at the generated Api Java file and can remove the currently implemented  required-checks then.

I suggest adding a CLI option for this:
I used the CLI-option useBeanValidationFeature for CXF server, but this is somewhat CXF-specific.
I think the better name would be performBeanValidation? Maybe we should also rename the CLI-flag for BeanValidation annotations "useBeanValidation" to useBeanValidationAnnotations to make it more clear only the annotations are generated?

So suggested CLI-flags would be:
* useBeanValidationAnnotations = generates the annotations only
* performBeanValidation = does the beanvalidation check automatically for the requests.
Maybe we could even add a separate flag to validate the response (performBeanValidationOnResponse)?

**Comment by jfiala** [comment]
@wing328 added PR #4177 with implementation of client-side beanvalidation.
